src/catch_unwind.rs

```diff
@@ -0,0 +1,57 @@
+use crate::{
+    middleware::{Middleware, Next},
+    response::IntoResponse,
+    Context, Response,
+};
+use futures::future::{BoxFuture, FutureExt, TryFutureExt};
+use http::status::StatusCode;
+use std::any::Any;
+
+/// A [`Middleware`] that will catch any panics from later middleware or handlers and return a
+/// response to the client.
+pub struct CatchUnwind {
+    f: Box<dyn Fn(Box<dyn Any + Send + 'static>) -> Response + Send + Sync>,
+}
+
+impl CatchUnwind {
+    /// Create a [`CatchUnwind`] which will respond with [`StatusCode::INTERNAL_SERVER_ERROR`] when
+    /// any panic is caught.
+    pub fn new() -> Self {
+        Self::with_response(|_| {
+            "Internal server error"
+                .with_status(StatusCode::INTERNAL_SERVER_ERROR)
+                .into_response()
+        })
+    }
+
+    /// Create a [`CatchUnwind`] with a custom function to generate the response, the function will
+    /// be passed the caught panic.
+    pub fn with_response(
+        response: impl Fn(Box<dyn Any + Send + 'static>) -> Response + Send + Sync + 'static,
+    ) -> Self {
+        Self {
+            f: Box::new(response),
+        }
+    }
+}
+
+impl<State: 'static> Middleware<State> for CatchUnwind {
+    fn handle<'a>(&'a self, cx: Context<State>, next: Next<'a, State>) -> BoxFuture<'a, Response> {
+        std::panic::AssertUnwindSafe(next.run(cx))
+            .catch_unwind()
+            .unwrap_or_else(move |err| (self.f)(err))
+            .boxed()
```

Comment:
This only returns 500 for a single request, right? My initial understanding was that _all_ requests would be terminated with 500, and then the `panic` would continue. But I guess the intended behavior here is different?

Reply:
Is there a need to terminate all requests? panic on the request path doesn't need to prevent other working code paths, as long as it's safely unwindable?

Reply:
Yes, this is the “I think my application will continue working if a single request panics” middleware. See the link in my other comment to the “fail-fast” version I haven’t finished yet.

Reply:
I was talking with the other Runtime maintainers yesterday, and if a future breaks inside of runtime somewhere (e.g. `runtime::spawn`), it'll default to terminating its thread, and we have no good way to restore that thread. There might not even be one tbh. I believe Tokio adopted a similar model recently too (on request of the Deno project).

If Tide starts catching panics rather than propagating them, we will start losing threads without having a way to spin them back up; which will severely limit throughput without failing the process entirely. Which seems like not a great situation to be in. I feel this would be particularly hard to debug when not intimately familiar with all moving parts up and down the stack (and even then it's not _easy_).

I guess what I'm saying is that this approach is making me feel quite nervous, and perhaps we should start with https://github.com/Nemo157/tide/compare/catch-unwind...propagate-unwind first?